File: golang-lib/broker.go
package aiomsg

// The broker: a single goroutine that owns the connection set and all routing,
// mirroring the Python reference's sender task + connections dict. Everything
// else talks to it by sending a command on its channel, so the connection map,
// round-robin cursor, send buffer, and in-flight ack table are touched from one
// goroutine — no locks on the hot path.

import (
	"context"
	"crypto/rand"
	"time"
)

const (
	resendTimeout = 5 * time.Second
	maxRetries    = 5
	recvBuffer    = 1024
	writerBuffer  = 256
)

// Message is a received message together with the identity of the sending peer.
type Message struct {
	Identity Identity
	Data     []byte
}

type cmdKind int

const (
	cmdSend cmdKind = iota
	cmdConnUp
	cmdConnDown
	cmdReceived
	cmdResend
)

// command is the single tagged union the broker consumes. Only the fields
// relevant to kind are set.
type command struct {
	kind        cmdKind
	identity    Identity
	hasIdentity bool
	data        []byte
	writer      chan []byte
	accept      chan bool
	env         envelope
	msgID       msgID
}

type bufferedMsg struct {
	identity    Identity
	hasIdentity bool
	data        []byte
}

type pending struct {
	identity    Identity
	hasIdentity bool
	data        []byte
	retries     int
	cancel      chan struct{}
}

type broker struct {
	ctx      context.Context
	sendMode SendMode
	delivery DeliveryGuarantee

	commands chan command
	recv     chan Message

	conns    map[Identity]chan []byte
	order    []Identity
	rrCursor int

	buffer  []bufferedMsg
	pending map[msgID]*pending
}

func newBroker(ctx context.Context, sendMode SendMode, delivery DeliveryGuarantee) *broker {
	return &broker{
		ctx:      ctx,
		sendMode: sendMode,
		delivery: delivery,
		commands: make(chan command),
		recv:     make(chan Message, recvBuffer),
		conns:    make(map[Identity]chan []byte),
		pending:  make(map[msgID]*pending),
	}
}

func (b *broker) run() {
	for {
		select {
		case <-b.ctx.Done():
			b.shutdown()
			return
		case cmd := <-b.commands:
			switch cmd.kind {
			case cmdSend:
				b.handleSend(cmd.identity, cmd.hasIdentity, cmd.data)
			case cmdConnUp:
				b.handleUp(cmd.identity, cmd.writer, cmd.accept)
			case cmdConnDown:
				b.handleDown(cmd.identity)
			case cmdReceived:
				b.handleReceived(cmd.identity, cmd.env)
			case cmdResend:
				b.handleResend(cmd.msgID)
			}
		}
	}
}

func (b *broker) shutdown() {
	for _, p := range b.pending {
		close(p.cancel)
	}
	for _, w := range b.conns {
		close(w)
	}
	close(b.recv)
}

func (b *broker) handleSend(id Identity, hasID bool, data []byte) {
	if len(b.conns) == 0 {
		b.buffer = append(b.buffer, bufferedMsg{id, hasID, data})
		return
	}
	b.transmit(id, hasID, data, maxRetries)
}

// transmit envelopes data and routes it, setting up resend tracking when the
// delivery guarantee calls for it.
func (b *broker) transmit(id Identity, hasID bool, data []byte, retries int) {
	atLeastOnce := b.delivery == AtLeastOnce && (hasID || b.sendMode == RoundRobin)
	if atLeastOnce {
		var mid msgID
		_, _ = rand.Read(mid[:])
		env := encodeDataReq(mid, data)

		cancel := make(chan struct{})
		b.startResendTimer(mid, cancel)
		b.pending[mid] = &pending{
			identity:    id,
			hasIdentity: hasID,
			data:        data,
			retries:     retries,
			cancel:      cancel,
		}
		b.route(id, hasID, env)
	} else {
		b.route(id, hasID, encodeData(data))
	}
}

func (b *broker) startResendTimer(mid msgID, cancel chan struct{}) {
	go func() {
		t := time.NewTimer(resendTimeout)
		defer t.Stop()
		select {
		case <-t.C:
			select {
			case b.commands <- command{kind: cmdResend, msgID: mid}:
			case <-b.ctx.Done():
			}
		case <-cancel:
		case <-b.ctx.Done():
		}
	}()
}

// route delivers an already-encoded envelope. Sends to writer channels are
// non-blocking: a full writer queue drops the message rather than stalling the
// whole broker (matches the reference).
func (b *broker) route(id Identity, hasID bool, env []byte) {
	if hasID {
		if ch, ok := b.conns[id]; ok {
			trySend(ch, env)
		}
		return
	}
	switch b.sendMode {
	case Publish:
		for _, ch := range b.conns {
			trySend(ch, env)
		}
	default: // RoundRobin
		if target, ok := b.nextRoundRobin(); ok {
			trySend(b.conns[target], env)
		}
	}
}

func trySend(ch chan []byte, env []byte) {
	select {
	case ch <- env:
	default: // writer queue full; drop
	}
}

func (b *broker) nextRoundRobin() (Identity, bool) {
	if len(b.order) == 0 {
		return Identity{}, false
	}
	id := b.order[b.rrCursor%len(b.order)]
	b.rrCursor++
	return id, true
}

func (b *broker) handleUp(id Identity, writer chan []byte, accept chan bool) {
	if _, exists := b.conns[id]; exists {
		accept <- false
		return
	}
	b.conns[id] = writer
	b.order = append(b.order, id)
	accept <- true

	if len(b.buffer) > 0 {
		buffered := b.buffer
		b.buffer = nil
		for _, m := range buffered {
			b.handleSend(m.identity, m.hasIdentity, m.data)
		}
	}
}

func (b *broker) handleDown(id Identity) {
	if ch, ok := b.conns[id]; ok {
		delete(b.conns, id)
		close(ch)
	}
	for i, x := range b.order {
		if x == id {
			b.order = append(b.order[:i], b.order[i+1:]...)
			break
		}
	}
	// rrCursor is always taken modulo len(order), so it needs no fixup.
}

func (b *broker) handleReceived(id Identity, env envelope) {
	switch env.Type {
	case typeData:
		b.deliver(Message{Identity: id, Data: env.Payload})
	case typeDataReq:
		// Deliver to the application, then acknowledge on the same connection.
		b.deliver(Message{Identity: id, Data: env.Payload})
		b.route(id, true, encodeAck(env.MsgID))
	case typeAck:
		if p, ok := b.pending[env.MsgID]; ok {
			delete(b.pending, env.MsgID)
			close(p.cancel)
		}
	}
}

func (b *broker) deliver(m Message) {
	select {
	case b.recv <- m:
	case <-b.ctx.Done():
	}
}

func (b *broker) handleResend(mid msgID) {
	p, ok := b.pending[mid]
	if !ok {
		return
	}
	delete(b.pending, mid)
	if p.retries == 0 {
		return
	}
	if len(b.conns) == 0 {
		b.buffer = append(b.buffer, bufferedMsg{p.identity, p.hasIdentity, p.data})
		return
	}
	b.transmit(p.identity, p.hasIdentity, p.data, p.retries-1)
}

File: rust-lib-async/src/broker.rs
//! The broker: the single owner of the connection set and all routing.
//!
//! Mirrors the Python reference's sender task + connections dict. Every other
//! part of the socket talks to the broker by sending it a [`Command`]; the
//! broker is the only place that touches the connection map, the round-robin
//! cursor, the send buffer, and the in-flight ack table. Keeping that state in
//! one task means no locks and no data races on the hot path.

use std::collections::{HashMap, VecDeque};
use std::time::Duration;

use bytes::Bytes;
use tokio::sync::{mpsc, oneshot};
use tokio::time::Instant;
use tracing::{debug, trace};

use crate::protocol::{self, Envelope, Identity, MsgId};
use crate::{DeliveryGuarantee, SendMode};

const RESEND_TIMEOUT: Duration = Duration::from_secs(5);
const MAX_RETRIES: u32 = 5;

/// Messages the broker processes. Every input to the broker — user sends,
/// connection lifecycle events, received frames — is one of these. Resend
/// timers are handled internally by the broker's run loop rather than
/// round-tripping through this channel (see `deadlines` on [`Broker`]).
pub(crate) enum Command {
    /// A user-level send. `identity` set means "to this peer only".
    Send {
        identity: Option<Identity>,
        data: Bytes,
    },
    /// A connection finished its handshake. `accept` reports back whether the
    /// broker kept it (false ⇒ duplicate identity, the connection should close).
    ConnectionUp {
        identity: Identity,
        writer: mpsc::UnboundedSender<Bytes>,
        accept: oneshot::Sender<bool>,
    },
    /// A connection ended.
    ConnectionDown { identity: Identity },
    /// A data envelope (DATA / DATA_REQ / ACK) arrived from a peer.
    Received {
        identity: Identity,
        envelope: Envelope,
    },
    /// Stop.
    Close,
}

/// Bookkeeping for one in-flight `AT_LEAST_ONCE` message awaiting its ACK.
struct Pending {
    identity: Option<Identity>,
    data: Bytes,
    retries: u32,
}

pub(crate) struct Broker {
    send_mode: SendMode,
    delivery: DeliveryGuarantee,
    /// Where received application payloads are delivered (to `Socket::recv`).
    recv_tx: mpsc::UnboundedSender<(Identity, Bytes)>,
    /// Active connections, keyed by peer identity, each value a channel into
    /// that connection's writer task.
    conns: HashMap<Identity, mpsc::UnboundedSender<Bytes>>,
    /// Connection identities in insertion order, for round-robin selection.
    order: Vec<Identity>,
    rr_cursor: usize,
    /// Messages sent while no peers were connected, flushed on the next connect.
    buffer: VecDeque<(Option<Identity>, Bytes)>,
    /// In-flight `AT_LEAST_ONCE` messages awaiting an ACK, keyed by msg_id.
    pending: HashMap<MsgId, Pending>,
    /// Resend deadlines in fire order. `RESEND_TIMEOUT` is constant and every
    /// entry is pushed at `Instant::now() + RESEND_TIMEOUT` from within this
    /// same single-threaded loop, so insertion order is already deadline
    /// order — a `VecDeque` is a min-heap here for free. An entry whose
    /// msg_id is no longer in `pending` when it reaches the front (because
    /// the ACK arrived, or a retry already replaced it with a fresh msg_id)
    /// is simply skipped.
    deadlines: VecDeque<(Instant, MsgId)>,
    /// Salt distinguishing this broker's msg_ids from any other broker's;
    /// combined with `msg_id_counter` to fill the 16-byte `MsgId`. Only needs
    /// to be unique per sender, not globally, so one random draw at startup
    /// plus a monotonic counter is enough — no per-message RNG call.
    msg_id_salt: [u8; 8],
    msg_id_counter: u64,
}

impl Broker {
    pub(crate) fn new(
        send_mode: SendMode,
        delivery: DeliveryGuarantee,
        recv_tx: mpsc::UnboundedSender<(Identity, Bytes)>,
    ) -> Self {
        let salt = uuid::Uuid::new_v4();
        let msg_id_salt: [u8; 8] = salt.as_bytes()[..8].try_into().unwrap();
        Broker {
            send_mode,
            delivery,
            recv_tx,
            conns: HashMap::new(),
            order: Vec::new(),
            rr_cursor: 0,
            buffer: VecDeque::new(),
            pending: HashMap::new(),
            deadlines: VecDeque::new(),
            msg_id_salt,
            msg_id_counter: 0,
        }
    }

    pub(crate) async fn run(mut self, mut cmd_rx: mpsc::UnboundedReceiver<Command>) {
        loop {
            // Next resend deadline, if any in-flight message is awaiting one.
            // Guarded by `!self.deadlines.is_empty()` below, so the dummy
            // `Instant::now()` fallback (needed only for the expression to
            // type-check) is never actually polled.
            let next_deadline = self
                .deadlines
                .front()
                .map_or_else(Instant::now, |(deadline, _)| *deadline);

            tokio::select! {
                cmd = cmd_rx.recv() => match cmd {
                    Some(Command::Send { identity, data }) => self.handle_send(identity, data),
                    Some(Command::ConnectionUp { identity, writer, accept }) => {
                        self.handle_up(identity, writer, accept)
                    }
                    Some(Command::ConnectionDown { identity }) => self.handle_down(identity),
                    Some(Command::Received { identity, envelope }) => {
                        self.handle_received(identity, envelope)
                    }
                    Some(Command::Close) | None => break,
                },
                () = tokio::time::sleep_until(next_deadline), if !self.deadlines.is_empty() => {
                    self.fire_due_resends();
                }
            }
        }
        // Returning drops `conns` (every connection writer sender) and
        // `recv_tx`, which ends the writer tasks and the receive stream.
    }

    fn handle_send(&mut self, identity: Option<Identity>, data: Bytes) {
        if self.conns.is_empty() {
            trace!("no peers; buffering message");
            self.buffer.push_back((identity, data));
            return;
        }
        self.transmit(identity, data, MAX_RETRIES);
    }

    /// Envelope `data` and route it, setting up resend tracking when the
    /// delivery guarantee calls for it.
    fn transmit(&mut self, identity: Option<Identity>, data: Bytes, retries: u32) {
        let at_least_once = self.delivery == DeliveryGuarantee::AtLeastOnce
            && (identity.is_some() || self.send_mode == SendMode::RoundRobin);

        if at_least_once {
            let msg_id = self.next_msg_id();
            let envelope = protocol::data_req(&msg_id, &data);

            // Schedule a resend; it fires unless an ACK removes `msg_id` from
            // `pending` first (see `fire_due_resends`).
            self.deadlines
                .push_back((Instant::now() + RESEND_TIMEOUT, msg_id));

            self.pending.insert(
                msg_id,
                Pending {
                    identity,
                    data,
                    retries,
                },
            );
            self.route(identity, envelope);
        } else {
            self.route(identity, protocol::data(&data));
        }
    }

    /// The next msg_id for an `AT_LEAST_ONCE` send: this broker's startup
    /// salt followed by a monotonic counter. Only needs to be unique among
    /// messages from this sender, which the counter guarantees without a
    /// per-message RNG call.
    fn next_msg_id(&mut self) -> MsgId {
        let counter = self.msg_id_counter;
        self.msg_id_counter = self.msg_id_counter.wrapping_add(1);
        let mut id = [0u8; 16];
        id[..8].copy_from_slice(&self.msg_id_salt);
        id[8..].copy_from_slice(&counter.to_be_bytes());
        id
    }

    /// Deliver an already-encoded envelope to the appropriate connection(s).
    fn route(&mut self, target: Option<Identity>, envelope: Bytes) {
        match target {
            Some(id) => match self.conns.get(&id) {
                Some(tx) => {
                    let _ = tx.send(envelope);
                }
                None => trace!("peer {} not connected; dropping", hex16(&id)),
            },
            None => match self.send_mode {
                SendMode::Publish => {
                    for tx in self.conns.values() {
                        let _ = tx.send(envelope.clone());
                    }
                }
                SendMode::RoundRobin => {
                    if let Some(id) = self.next_round_robin() {
                        if let Some(tx) = self.conns.get(&id) {
                            let _ = tx.send(envelope);
                        }
                    }
                }
            },
        }
    }

    fn next_round_robin(&mut self) -> Option<Identity> {
        if self.order.is_empty() {
            return None;
        }
        let id = self.order[self.rr_cursor % self.order.len()];
        self.rr_cursor = self.rr_cursor.wrapping_add(1);
        Some(id)
    }

    fn handle_up(
        &mut self,
        identity: Identity,
        writer: mpsc::UnboundedSender<Bytes>,
        accept: oneshot::Sender<bool>,
    ) {
        if self.conns.contains_key(&identity) {
            debug!("duplicate identity {}; rejecting", hex16(&identity));
            let _ = accept.send(false);
            return;
        }
        self.conns.insert(identity, writer);
        self.order.push(identity);
        let _ = accept.send(true);

        // First peer available: flush anything buffered while disconnected.
        if !self.buffer.is_empty() {
            let buffered: Vec<_> = self.buffer.drain(..).collect();
            for (id, data) in buffered {
                self.handle_send(id, data);
            }
        }
    }

    fn handle_down(&mut self, identity: Identity) {
        self.conns.remove(&identity);
        if let Some(pos) = self.order.iter().position(|x| *x == identity) {
            self.order.remove(pos);
        }
        // rr_cursor is always taken modulo order.len(), so it needs no fixup.
    }

    fn handle_received(&mut self, identity: Identity, envelope: Envelope) {
        match envelope {
            Envelope::Data { payload } => {
                let _ = self.recv_tx.send((identity, payload));
            }
            Envelope::DataReq { msg_id, payload } => {
                // Deliver to the application first, then acknowledge on the same
                // connection the request arrived on.
                let _ = self.recv_tx.send((identity, payload));
                self.route(Some(identity), protocol::ack(&msg_id));
            }
            Envelope::Ack { msg_id } => {
                // Just drop the bookkeeping; the matching `deadlines` entry
                // is skipped as stale when it reaches the front (see
                // `fire_due_resends`).
                self.pending.remove(&msg_id);
            }
            // HELLO / HEARTBEAT are handled in the connection layer and never
            // reach the broker.
            Envelope::Hello { .. } | Envelope::Heartbeat => {}
        }
    }

    /// Pop and process every deadline that has already passed. Called once
    /// per timer fire; loops in case several deadlines are due in the same
    /// tick (e.g. after a burst of `AT_LEAST_ONCE` sends).
    fn fire_due_resends(&mut self) {
        let now = Instant::now();
        while let Some(&(deadline, msg_id)) = self.deadlines.front() {
            if deadline > now {
                break;
            }
            self.deadlines.pop_front();
            self.fire_resend(msg_id);
        }
    }

    fn fire_resend(&mut self, msg_id: MsgId) {
        // Absent means the ACK already arrived (or a previous retry already
        // superseded this msg_id) — nothing to do.
        let Some(pending) = self.pending.remove(&msg_id) else {
            return;
        };
        if pending.retries == 0 {
            debug!("max retries reached; dropping message");
            return;
        }
        if self.conns.is_empty() {
            self.buffer.push_back((pending.identity, pending.data));
            return;
        }
        self.transmit(pending.identity, pending.data, pending.retries - 1);
    }
}

/// Short hex of an identity for log lines.
fn hex16(id: &Identity) -> String {
    id.iter().map(|b| format!("{b:02x}")).collect()
}
